File: app/src/main/java/jp/hazuki/yuzubrowser/utils/view/swipebutton/SwipeController.java
package jp.hazuki.yuzubrowser.utils.view.swipebutton;

import android.content.Context;
import android.view.GestureDetector;
import android.view.MotionEvent;

public class SwipeController {
    public static final int SWIPE_CANCEL = 0;
    public static final int SWIPE_RIGHT = 0x01;
    public static final int SWIPE_LEFT = 0x02;
    public static final int SWIPE_UP = 0x04;
    public static final int SWIPE_DOWN = 0x08;
    public static final int SWIPE_X_Y = SWIPE_RIGHT | SWIPE_LEFT | SWIPE_UP | SWIPE_DOWN;
    public static final int SWIPE_PRESS = 0x10;
    public static final int SWIPE_LPRESS = 0x20;

    private int mWhatNo = SWIPE_PRESS;
    private int mSense;

    private final GestureDetector mDetector;

    public interface OnChangeListener {
        boolean onEventOutSide();

        boolean onEventCancel();

        boolean onEventActionUp(int whatNo);

        boolean onEventActionDown();

        void onChangeState(int whatNo);

        void onLongPress();
    }

    public void setOnChangeListener(OnChangeListener l) {
        mListener = l;
    }

    private OnChangeListener mListener = null;

    public SwipeController(Context context) {
        mDetector = new GestureDetector(context, new GestureDetector.SimpleOnGestureListener() {
            @Override
            public boolean onScroll(MotionEvent e1, MotionEvent e2, float distanceX, float distanceY) {
                int rangeX = (int) (e2.getRawX() - e1.getRawX());
                int rangeY = (int) (e2.getRawY() - e1.getRawY());

                int check = 0;
                if (rangeX > mSense)
                    check |= SWIPE_RIGHT;
                else if (rangeX < -mSense)
                    check |= SWIPE_LEFT;
                if (rangeY > mSense)
                    check |= SWIPE_DOWN;
                else if (rangeY < -mSense)
                    check |= SWIPE_UP;

                if (check == 0) {
                    if (mWhatNo != SWIPE_CANCEL) {
                        mWhatNo = SWIPE_CANCEL;
                        if (mListener != null) mListener.onChangeState(SWIPE_CANCEL);
                    }
                } else {
                    switch (check & SWIPE_X_Y) {
                        case SWIPE_RIGHT:
                            callListener(SWIPE_RIGHT);
                            break;
                        case SWIPE_LEFT:
                            callListener(SWIPE_LEFT);
                            break;
                        case SWIPE_DOWN:
                            callListener(SWIPE_DOWN);
                            break;
                        case SWIPE_UP:
                            callListener(SWIPE_UP);
                            break;
                        case SWIPE_RIGHT | SWIPE_DOWN:
                            callListener((rangeX > rangeY) ? SWIPE_RIGHT : SWIPE_DOWN);
                            break;
                        case SWIPE_RIGHT | SWIPE_UP:
                            callListener((rangeX > -rangeY) ? SWIPE_RIGHT : SWIPE_UP);
                            break;
                        case SWIPE_LEFT | SWIPE_DOWN:
                            callListener((-rangeX > rangeY) ? SWIPE_LEFT : SWIPE_DOWN);
                            break;
                        case SWIPE_LEFT | SWIPE_UP:
                            callListener((-rangeX > -rangeY) ? SWIPE_LEFT : SWIPE_UP);
                            break;
                    }
                }

                return super.onScroll(e1, e2, distanceX, distanceY);
            }

            private void callListener(int what) {
                if (mWhatNo != what) {
                    mWhatNo = what;
                    if (mListener != null)
                        mListener.onChangeState(what);
                }
            }

            @Override
            public boolean onSingleTapUp(MotionEvent e) {
                return super.onSingleTapUp(e);
            }

            @Override
            public void onLongPress(MotionEvent e) {
                if (mWhatNo == SWIPE_PRESS) {
                    mWhatNo = SWIPE_LPRESS;
                    onEventLongPress();
                    if (mListener != null) {
                        mListener.onChangeState(mWhatNo);
                        mListener.onLongPress();
                    }
                }
            }
        });
    }

    public boolean onTouchEvent(MotionEvent event) {
        mDetector.onTouchEvent(event);
        switch (event.getActionMasked()) {
            case MotionEvent.ACTION_DOWN:
                onEventActionDown();
                if (mListener != null) return mListener.onEventActionDown();
                break;
            case MotionEvent.ACTION_UP:
                int tmp_no = mWhatNo;
                mWhatNo = SWIPE_PRESS;
                boolean ret = false;
                onEventActionUp(tmp_no);
                if (mListener != null) ret = mListener.onEventActionUp(tmp_no);
                return ret;
            case MotionEvent.ACTION_CANCEL:
                onEventCancel();
                if (mListener != null) return mListener.onEventCancel();
                break;
            case MotionEvent.ACTION_OUTSIDE:
                onEventOutSide();
                if (mListener != null) return mListener.onEventOutSide();
                break;
        }
        return false;
    }

    public void notifyChangeState() {
        if (mListener != null) mListener.onChangeState(mWhatNo);
    }

    public void setToDefault() {
        mWhatNo = SWIPE_PRESS;
        mListener.onChangeState(mWhatNo);
    }

    public void setSense(int sense) {
        mSense = sense;
    }

    protected int getCurrentWhatNo() {
        return mWhatNo;
    }

    protected void onEventActionDown() {
    }

    protected void onEventActionUp(int whatNo) {
    }

    protected void onEventLongPress() {
    }

    protected void onEventCancel() {
    }

    protected void onEventOutSide() {
    }
}

File: module/ui/src/main/java/jp/hazuki/yuzubrowser/ui/widget/swipebutton/SwipeController.kt
/*
 * Copyright (C) 2017-2019 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.ui.widget.swipebutton

import android.content.Context
import android.view.GestureDetector
import android.view.MotionEvent

open class SwipeController(context: Context) {

    protected var currentWhatNo = SWIPE_PRESS
        private set
    private var sense: Int = 0

    private val detector: GestureDetector

    private var listener: OnChangeListener? = null

    interface OnChangeListener {
        fun onEventOutSide(): Boolean

        fun onEventCancel(): Boolean

        fun onEventActionUp(whatNo: Int): Boolean

        fun onEventActionDown(): Boolean

        fun onChangeState(whatNo: Int)

        fun onLongPress()
    }

    fun setOnChangeListener(l: OnChangeListener) {
        listener = l
    }

    init {
        detector = GestureDetector(context, object : GestureDetector.SimpleOnGestureListener() {
            override fun onScroll(e1: MotionEvent, e2: MotionEvent, distanceX: Float, distanceY: Float): Boolean {
                val rangeX = (e2.rawX - e1.rawX).toInt()
                val rangeY = (e2.rawY - e1.rawY).toInt()

                var check = 0
                if (rangeX > sense)
                    check = check or SWIPE_RIGHT
                else if (rangeX < -sense)
                    check = check or SWIPE_LEFT
                if (rangeY > sense)
                    check = check or SWIPE_DOWN
                else if (rangeY < -sense)
                    check = check or SWIPE_UP

                if (check == 0) {
                    if (currentWhatNo != SWIPE_CANCEL) {
                        currentWhatNo = SWIPE_CANCEL
                        listener?.onChangeState(SWIPE_CANCEL)
                    }
                } else {
                    when (check and SWIPE_X_Y) {
                        SWIPE_RIGHT -> callListener(SWIPE_RIGHT)
                        SWIPE_LEFT -> callListener(SWIPE_LEFT)
                        SWIPE_DOWN -> callListener(SWIPE_DOWN)
                        SWIPE_UP -> callListener(SWIPE_UP)
                        SWIPE_RIGHT or SWIPE_DOWN -> callListener(if (rangeX > rangeY) SWIPE_RIGHT else SWIPE_DOWN)
                        SWIPE_RIGHT or SWIPE_UP -> callListener(if (rangeX > -rangeY) SWIPE_RIGHT else SWIPE_UP)
                        SWIPE_LEFT or SWIPE_DOWN -> callListener(if (-rangeX > rangeY) SWIPE_LEFT else SWIPE_DOWN)
                        SWIPE_LEFT or SWIPE_UP -> callListener(if (-rangeX > -rangeY) SWIPE_LEFT else SWIPE_UP)
                    }
                }

                return super.onScroll(e1, e2, distanceX, distanceY)
            }

            private fun callListener(what: Int) {
                if (currentWhatNo != what) {
                    currentWhatNo = what
                    listener?.onChangeState(what)
                }
            }

            override fun onLongPress(e: MotionEvent) {
                if (currentWhatNo == SWIPE_PRESS) {
                    currentWhatNo = SWIPE_LPRESS
                    onEventLongPress()
                    listener?.run {
                        onChangeState(currentWhatNo)
                        onLongPress()
                    }
                }
            }
        })
    }

    fun onTouchEvent(event: MotionEvent): Boolean {
        detector.onTouchEvent(event)
        when (event.actionMasked) {
            MotionEvent.ACTION_DOWN -> {
                onEventActionDown()
                listener?.let { return it.onEventActionDown() }
            }
            MotionEvent.ACTION_UP -> {
                val tmpNo = currentWhatNo
                currentWhatNo = SWIPE_PRESS
                var ret = false
                onEventActionUp(tmpNo)
                listener?.let { ret = it.onEventActionUp(tmpNo) }
                return ret
            }
            MotionEvent.ACTION_CANCEL -> {
                onEventCancel()
                listener?.let { return it.onEventCancel() }
            }
            MotionEvent.ACTION_OUTSIDE -> {
                onEventOutSide()
                listener?.let { return it.onEventOutSide() }
            }
        }
        return false
    }

    fun notifyChangeState() {
        listener?.onChangeState(currentWhatNo)
    }

    fun setToDefault() {
        currentWhatNo = SWIPE_PRESS
        listener?.onChangeState(currentWhatNo)
    }

    fun setSense(sense: Int) {
        this.sense = sense
    }

    protected open fun onEventActionDown() {}

    protected open fun onEventActionUp(whatNo: Int) {}

    protected open fun onEventLongPress() {}

    protected open fun onEventCancel() {}

    protected open fun onEventOutSide() {}

    companion object {
        const val SWIPE_CANCEL = 0x00
        const val SWIPE_RIGHT = 0x01
        const val SWIPE_LEFT = 0x02
        const val SWIPE_UP = 0x04
        const val SWIPE_DOWN = 0x08
        const val SWIPE_X_Y = SWIPE_RIGHT or SWIPE_LEFT or SWIPE_UP or SWIPE_DOWN
        const val SWIPE_PRESS = 0x10
        const val SWIPE_LPRESS = 0x20
    }
}
